### PR TITLE
[Spec Decoding] Support Qwen3 DSpark in Stage1

### PR DESCRIPTION
**Pull request by lingebeng**
## Benchmark

- **Hardware:** TPU v7x-8; TP8×DP1 and TP4×DP2.
- **Target:** `Qwen/Qwen3-8B`
- **Draft:** `deepseek-ai/dspark_qwen3_8b_block7`
- **Serving:** Overlap enabled, BF16, FA, memory fraction `0.65`, page size `64`, chunked prefill `2048`, maximum running requests `128`, radix cache disabled.

### Speed Benchmark

Dataset: `random`; input/output lengths: **1024/4096 tokens**; range ratio: `1.0`; requests: **3×concurrency**; request rate: `inf`. Recorded cache hit rate: **0**.

Output throughput in **tokens/s**:

| Concurrency | TP8×DP1 Baseline | TP8×DP1 DSpark | Speedup | TP4×DP2 Baseline | TP4×DP2 DSpark | Speedup |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 106.8 | 537.2 | 5.03× | 102.5 | 519.1 | 5.06× |
| 2 | 211.2 | 1,022.9 | 4.84× | 204.0 | 874.2 | 4.28× |
| 4 | 412.4 | 1,964.5 | 4.76× | 404.1 | 1,958.3 | 4.85× |
| 8 | 810.1 | 3,222.6 | 3.98× | 793.3 | 3,283.1 | 4.14× |
| 16 | 1,505.7 | 6,506.9 | 4.32× | 1,559.3 | 7,754.5 | 4.97× |
| 32 | 2,838.1 | 9,682.3 | 3.41× | 2,945.0 | 10,544.8 | 3.58× |
| 64| 3,829.9 | 12,185.0 | 3.18× | 4,517.5 | 15,134.9 | 3.35× |
| 128| 4,365.9 | 13,899.2 | 3.18× | 5,577.7 | 17,656.4 | 3.17× |

### Accuracy Benchmark

- **Dataset:** GSM8K, 500 samples, using DeepSpec’s data selection and prompt format.
- **Seed:** `980406`
- **Generation:** Temperature `0`, thinking disabled, maximum output `2048` tokens, concurrency `16`.
- **Scoring:** Custom numeric exact match with answer-format normalization.
- This differs from the paper’s **temperature=1** evaluation protocol.

| Configuration | Baseline Accuracy | DSpark Accuracy | DSpark Acceptance Length |
|---|---:|---:|---:|
| TP8×DP1 | 92.8% (464/500) | 93.2% (466/500) | 5.488 |
| TP4×DP2 | 92.8% (464/500) | 93.0% (465/500) | 5.473 |

## Repro

Use `feat/dspark-stage1` at commit `0d0271d0f76376d09bf4b12f663240a7f4e8c989`

Launch each configuration on TPU v7x-8:

```bash
DP=1         # 1 or 2
MODE=dspark  # dspark or baseline

SPEC_ARGS=()
if [[ "$MODE" == "dspark" ]]; then
  SPEC_ARGS=(
    --speculative-algorithm DSPARK
    --speculative-draft-model-path deepseek-ai/dspark_qwen3_8b_block7
    --speculative-num-draft-tokens 8
    --speculative-num-steps 1
    --speculative-eagle-topk 1
  )
fi

python -m sgl_jax.launch_server \
  --model-path Qwen/Qwen3-8B \
  "${SPEC_ARGS[@]}" \
  --tp-size 8 \
  --dp-size "$DP" \
  --dtype bfloat16 \
  --attention-backend fa \
  --mem-fraction-static 0.65 \
  --page-size 64 \
  --chunked-prefill-size 2048 \
  --max-running-requests 128 \
  --trust-remote-code \
  --disable-radix-cache \
  --grammar-backend none \
  --random-seed 980406 \
  --host 0.0.0.0 \
  --port 30000
```

On this branch, `--tp-size` specifies the total mesh size: `8/1` produces TP8×DP1, while `8/2` produces TP4×DP2. Overlap remains enabled by default, and `DSPARK` automatically enables sampling from the anchor.

After the server becomes ready, run the speed benchmark:

```bash
for C in 1 2 4 8 16 32 64 128; do
  OUT="results/dp${DP}/${MODE}/bs_${C}"
  mkdir -p "$OUT"

  python -m sgl_jax.bench_serving \
    --backend sgl-jax \
    --host 127.0.0.1 \
    --port 30000 \
    --dataset-name random \
    --random-input-len 1024 \
    --random-output-len 4096 \
    --random-range-ratio 1.0 \
    --num-prompts "$((3 * C))" \
    --max-concurrency "$C" \
    --request-rate inf \
    --seed 980406 \
    --output-file "$OUT/result.jsonl"
done
```

For GSM8K, shuffle the official test split with `random.Random(980406)` and select the first 500 samples. Apply the Qwen3 chat template with `enable_thinking=False`, appending:

```text
Please reason step by step, and put your final answer within \boxed{}.
```

Send requests to `/generate` with temperature `0`, `max_new_tokens=2048`, and concurrency `16`. Score the final boxed numeric answer against the GSM8K gold answer using the same formatting normalization for all configurations. Measure acceptance length from the difference in server acceptance counters before and after evaluation.

This accuracy procedure uses DeepSpec’s data selection and prompt format, but differs from the paper’s temperature-1 sampling protocol.

### CODE REVIEW

File: docs/features/speculative_decoding.md [hunk 1]
```diff
@@ -150,10 +150,56 @@ accepted-context KV materialization retain the full verification width. Both
...
+For `deepseek-ai/dspark_qwen3_8b_block7`:
+
+```bash
+--speculative-algorithm DSPARK \
+--speculative-draft-model-path deepseek-ai/dspark_qwen3_8b_block7 \
+--speculative-num-draft-tokens 8 \
+--speculative-num-steps 1 \
+--speculative-eagle-topk 1 \
+--grammar-backend none \
+--disable-overlap-schedule
```
Comment (cjx0709) [inline docs/features/speculative_decoding.md]: now in the benchmark, this cli arg not happend

Comment (lingebeng) [reply docs/features/speculative_decoding.md]: OK

**Review by cjx0709 — APPROVED** [review]
LGTM
